crates/prover-tracer/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+pub enum TracingOutput {
+    #[default]
+    Stdout,
+    Stderr,
+    File(PathBuf),
+    Otlp,
+}
+
+impl<'de> Deserialize<'de> for TracingOutput {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        // If the string is not recognized to be either `stdout` or `stderr`,
+        // it is assumed to be a file path.
+        match s.as_str() {
+            "stdout" => Ok(TracingOutput::Stdout),
+            "stderr" => Ok(TracingOutput::Stderr),
+            "otlp" => Ok(TracingOutput::Otlp),
+            _ => Ok(TracingOutput::File(PathBuf::from(s))),
+        }
+    }
+}
+
+impl TracingOutput {
+    /// Get a [`BoxMakeWriter`] for the trace output.
+    ///
+    /// This can be used to plug the trace output into the tracing subscriber.
+    pub fn as_make_writer(&self) -> BoxMakeWriter {
+        match self {
+            TracingOutput::Stdout => BoxMakeWriter::new(std::io::stdout),
+            TracingOutput::Stderr => BoxMakeWriter::new(std::io::stderr),
+            // For OTLP, output traces also to stdout
+            TracingOutput::Otlp => BoxMakeWriter::new(std::io::stdout),
+            TracingOutput::File(path) => {
+                let appender = tracing_appender::rolling::never(".", path);
+                BoxMakeWriter::new(appender)
+            }
+        }
+    }
+}
+
+pub fn setup_tracing(config: &TracingConfig, version: &str) -> anyhow::Result<()> {
+    let writer = config.outputs.first().cloned().unwrap_or_default();
+
+    let mut layers = Vec::new();
+
+    // Setup instrumentation if both otlp agent url and
+    // otlp service name are provided as arguments
+    if config.outputs.contains(&TracingOutput::Otlp) {
+        if let (Some(otlp_agent), Some(otlp_service_name)) =
+            (&config.otlp_agent, &config.otlp_service_name)
+        {
+            let resources = build_resources(otlp_service_name, version);
+            let otlp_exporter = opentelemetry_otlp::SpanExporter::builder()
+                .with_tonic()
+                .with_endpoint(otlp_agent)
+                .build()?;
+
+            let batch_processor_config = BatchConfigBuilder::default()
+                .with_scheduled_delay(match std::env::var("OTLP_BATCH_SCHEDULED_DELAY") {
+                    Ok(v) => Duration::from_millis(v.parse::<u64>().unwrap_or(5_000)),
+                    _ => Duration::from_millis(5_000),
+                })
+                .with_max_queue_size(match std::env::var("OTLP_BATCH_MAX_QUEUE_SIZE") {
+                    Ok(v) => v.parse::<usize>().unwrap_or(2048),
+                    _ => 2048,
+                })
+                .with_max_export_batch_size(
+                    match std::env::var("OTLP_BATCH_MAX_EXPORTER_BATCH_SIZE") {
+                        Ok(v) => v.parse::<usize>().unwrap_or(512),
+                        _ => 512,
+                    },
+                );
```

Comment:
Here too, the various `parse` calls could return an error rather than silently accept a malformed input and fabricate a value out of thin air. Same further down below this comment.